### PR TITLE
new route for weekly checkout and split out bundle

### PR DESCRIPTION
**Pull request by johnduffell**
## Why are you doing this?

This creates the new route and skeleton bundle for the weekly checkout.
This was mostly mechanical however I couldn't bring myself to copy and paste all the bundle without moving the payment related stuff somewhere shared.  However I went on a learning journey on the way round so it prob needs a bit more cleanup after that.
It would be good to get a look to make sure I'm on the right track though!

 I didn't do a URL parameter in the end because getting it to work without logging in would be a bit tricky

[**Trello Card**](https://trello.com/c/DlpF9cS6/2332-guardian-weekly-checkout-url-and-page-skeleton)

## Changes

* add route
* add controller
* add page template
* copy bundle and make sure it only duplicates the non common stuff

## Screenshots

image

### CODE REVIEW

File: support-frontend/app/controllers/WeeklySubscription.scala [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+  implicit val a: AssetsResolver = assets
+
+
+  def displayForm(): Action[AnyContent] =
+    (PrivateAction andThen authAction andThen authenticatedAction(subscriptionsClientId)).async { implicit request =>
+      implicit val settings: AllSettings = settingsProvider.getAllSettings()
+      identityService.getUser(request.user).fold(
+        error => {
+          SafeLogger.error(scrub"Failed to display paper subscriptions form for ${request.user.id} due to error from identityService: $error")
```
Comment (rupertbates) [inline support-frontend/app/controllers/WeeklySubscription.scala]: Should change this log message to reflect the fact that it's Guardian Weekly

File: support-frontend/app/views/main.scala.html [hunk 2]
```diff
@@ -117,7 +117,7 @@
       @ssrContent.content
     </div>
   }
-
+I
```
Comment (rupertbates) [inline support-frontend/app/views/main.scala.html]: ?

File: support-frontend/assets/helpers/paymentProviders.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  price: number,
+  currency: IsoCurrency,
+  email: string,
+) {
+  loadStripe()
+    .then(() => setupStripeCheckout(onAuthorised, 'REGULAR', currency, isTestUser))
+    .then(stripe => openDialogBox(stripe, price, email));
+}
+
+const countrySupportsDirectDebit = (country: ?IsoCountry): boolean => country !== null && country === 'GB';
```
Comment (rupertbates) [inline support-frontend/assets/helpers/paymentProviders.js]: Don't worry for this PR but we should move this function to a top level helper module rather than copying it in every checkout

File: support-frontend/assets/helpers/user/getUser.js [hunk 4]
```diff
@@ -1,13 +1,13 @@
 // @flow
 import { type Option } from 'helpers/types/option';
 
-export type User = {|
+export type GetUser = {|
```
Comment (rupertbates) [inline support-frontend/assets/helpers/user/getUser.js]: Can we keep this type as `User` please as that is what it represents.

File: support-frontend/assets/helpers/dateConversions.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+// @flow
+
+const formatMachineDate = (date: Date) =>
```
Comment (rupertbates) [inline support-frontend/assets/helpers/dateConversions.js]: 👍 good to move these to their own module

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/paymentProviders.js [hunk 6]
```diff
@@ -1,34 +1,42 @@
...
   setSubmissionError,
   type State,
 } from '../paperSubscriptionCheckoutReducer';
 import { DirectDebit, Stripe } from 'helpers/paymentMethods';
 
+import {
+  getBillingAddress,
+  getDeliveryAddress,
+} from '../paperSubscriptionCheckoutReducer';
+import type {IsoCurrency} from "../../../helpers/internationalisation/currency";
```
Comment (rupertbates) [inline support-frontend/assets/pages/paper-subscription-checkout/helpers/paymentProviders.js]: This can just be `helpers/internationalisation/currency`

You can get Intellij to do this by default by checking the 'Use paths relative to project, resource or sources roots' checkbox in Settings | Editor | Code Style | JavaScript | Imports

File: support-frontend/assets/helpers/paymentProviders.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+// @flow
+
+import { loadStripe, openDialogBox, setupStripeCheckout } from 'helpers/paymentIntegrations/stripeCheckout';
+import { type IsoCountry } from 'helpers/internationalisation/country';
+import { type PaymentAuthorisation } from 'helpers/paymentIntegrations/readerRevenueApis';
+
+import type {IsoCurrency} from "./internationalisation/currency";
+
+function showStripe(
```
Comment (rupertbates) [inline support-frontend/assets/helpers/paymentProviders.js]: 🥇 for moving these into their own module

File: support-frontend/assets/pages/weekly-subscription-checkout/components/checkoutForm.jsx [hunk 8]
```diff
@@ -0,0 +1,223 @@
+// @flow
+
+// ----- Imports ----- //
```
Comment (rupertbates) [inline support-frontend/assets/pages/weekly-subscription-checkout/components/checkoutForm.jsx]: I think it might be worth calling this `weeklyCheckoutForm.jsx` as we currently have 3 other jsx files and one js file called checkout! We could then rename the equivalent files in paper and digital, but that could be a separate PR

File: support-frontend/assets/pages/weekly-subscription-checkout/components/checkoutForm.jsx [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+            />
+          }
+          title="weekly holding page"
+          description="placeholder description"
+          productPrice={{price: 99.89, currency: 'GBP'}}
+          promotion={undefined}
+          dataList={[
+          ]}
+          billingPeriod="Monthly"
+          changeSubscription={routes.paperSubscriptionProductChoices}
```
Comment (rupertbates) [inline support-frontend/assets/pages/weekly-subscription-checkout/components/checkoutForm.jsx]: This probably needs to change? But if this is just a placeholder at this stage then don't worry.

File: support-frontend/assets/pages/weekly-subscription-checkout/weeklySubscriptionCheckout.jsx [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+const content = (
+  <Provider store={store}>
+    <Page
+      header={<Header displayNavigation={false} />}
+      footer={
+        <Footer>
+          <SubscriptionTermsPrivacy subscriptionProduct="Paper" />
+          <CustomerService
+            selectedCountryGroup={countryGroupId}
+            subscriptionProduct="Paper"
```
Comment (rupertbates) [inline support-frontend/assets/pages/weekly-subscription-checkout/weeklySubscriptionCheckout.jsx]: `GuardianWeekly`

File: support-frontend/assets/pages/weekly-subscription-checkout/weeklySubscriptionCheckout.jsx [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+import Header from 'components/headers/header/header';
+import Footer from 'components/footer/footer';
+import CustomerService from 'components/customerService/customerService';
+import SubscriptionTermsPrivacy from 'components/legal/subscriptionTermsPrivacy/subscriptionTermsPrivacy';
+import SubscriptionFaq from 'components/subscriptionFaq/subscriptionFaq';
+import 'stylesheets/skeleton/skeleton.scss';
+
+import { initReducer } from './weeklySubscriptionCheckoutReducer';
+import CheckoutStage from './stage';
+// import { type PaperFulfilmentOptions } from 'helpers/productPrice/fulfilmentOptions';
```
Comment (rupertbates) [inline support-frontend/assets/pages/weekly-subscription-checkout/weeklySubscriptionCheckout.jsx]: get rid of commented imports

File: support-frontend/conf/routes [hunk 12]
```diff
@@ -97,6 +97,7 @@ GET  /subscribe/digital/checkout/thankyou-existing  controllers.DigitalSubscript
 
 GET  /weekly                            controllers.Subscriptions.weeklyGeoRedirect()
 GET  /subscribe/weekly                            controllers.Subscriptions.weeklyGeoRedirect()
+GET  /subscribe/weekly/checkout                            controllers.WeeklySubscription.displayForm()
```
Comment (rupertbates) [inline support-frontend/conf/routes]: add a query param to this route, eg. `enableCheckout` and if it's not present in the url redirect to the old version of the checkout on subscribe.

File: support-frontend/app/controllers/WeeklySubscription.scala [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  fontLoaderBundle: Either[RefPath, StyleContent]
+)(implicit val ec: ExecutionContext) extends AbstractController(components) with GeoRedirect with Circe with CanonicalLinks with SettingsSurrogateKeySyntax {
+
+  import actionRefiners._
+
+  implicit val a: AssetsResolver = assets
+
+
+  def displayForm(): Action[AnyContent] =
+    (PrivateAction andThen authAction andThen authenticatedAction(subscriptionsClientId)).async { implicit request =>
```
Comment (johnduffell) [inline support-frontend/app/controllers/WeeklySubscription.scala]: oh yeah I also added google auth to it to prevent it being discovered by search engines before I remembered we were going to use a URL parameter originally.
The reason we were suggesting that is so we didn't forget to remove it before going live.  So I will add something obvious in the URL before shipping this.

File: support-frontend/app/views/testUsers.scala.html [hunk 14]
```diff
@@ -10,7 +10,7 @@
         <title>Test Users</title>
     </head>
     <body>
-        <h1>New Test User key: @key</h1>
+        <h1>New Test GetUser key: @key</h1>
```
Comment (johnduffell) [inline support-frontend/app/views/testUsers.scala.html]: intellij isn't so brilliant at renaming variables in JS code... 🙈

File: support-frontend/app/wiring/Controllers.scala [hunk 15]
```diff
@@ -77,6 +77,22 @@ trait Controllers {
     fontLoader
   )
 
+  lazy val weeklyController = new WeeklySubscription(
+    authAction,
```
Comment (johnduffell) [inline support-frontend/app/wiring/Controllers.scala]: this is a lot of parameters for one class! surely every method can't need all of them 🤔

File: support-frontend/assets/helpers/dateConversions.js (hunk 5 above)
Comment (johnduffell) [reply support-frontend/assets/helpers/dateConversions.js]: thanks I'm glad it turned out OK!

File: support-frontend/assets/helpers/paymentProviders.js (hunk 7 above)
Comment (johnduffell) [reply support-frontend/assets/helpers/paymentProviders.js]: do the parameters look sensible? there were a lot of paper specific "super objects" being passed around, and I ended up doing quite a bit of unpacking hence the parameter list looks a bit long.  Also I ended up going around the houses learning how the architecture fits together, so I have ended up simplified but increased the quantity of the params in the product specific versions of `paymentProviders.js`.
They could probably be refactored back to the paper/weekly specific objects if that's better.

File: support-frontend/assets/helpers/user/getUser.js (hunk 4 above)
Comment (johnduffell) [reply support-frontend/assets/helpers/user/getUser.js]: yes of course, another intellij mistake there that I hadn't self reviewed yet

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/paymentProviders.js (hunk 6 above)
Comment (johnduffell) [reply support-frontend/assets/pages/paper-subscription-checkout/helpers/paymentProviders.js]: now this is the kind of comment I like in my PRs! 💪 power up!

Comment (johnduffell) [reply support-frontend/assets/pages/paper-subscription-checkout/helpers/paymentProviders.js]: erm actually this file the imports are really messed up! I'll tidy.  Do we have a specific ordering e.g. alphabetical?

File: support-frontend/app/views/testUsers.scala.html (hunk 14 above)
Comment (rupertbates) [reply support-frontend/app/views/testUsers.scala.html]: No it's very poor

File: support-frontend/assets/helpers/paymentProviders.js (hunk 7 above)
Comment (rupertbates) [reply support-frontend/assets/helpers/paymentProviders.js]: Yes I think they're fine.

File: support-frontend/assets/pages/weekly-subscription-checkout/stage.jsx [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+import { type CountryGroupId } from 'helpers/internationalisation/countryGroup';
+
+import ProgressMessage from 'components/progressMessage/progressMessage';
+
+import { type Stage, type State } from './weeklySubscriptionCheckoutReducer';
+
+import CheckoutForm from './components/checkoutForm';
+import ReturnSection from 'pages/paper-subscription-checkout/components/thankYou/returnSection';
+import ThankYouContent from 'pages/paper-subscription-checkout/components/thankYou/thankYou';
```
Comment (johnduffell) [inline support-frontend/assets/pages/weekly-subscription-checkout/stage.jsx]: these thank you components might have commonality that needs extracting. That could be a task for the next PR

File: support-frontend/conf/routes (hunk 12 above)
Comment (johnduffell) [reply support-frontend/conf/routes]: couldn't work out how to do a query param quickly and I'm going now, so I've done two separate routes for now. but we can change it in this/another PR later.

**Comment by johnduffell** [comment]
I think most comments are addressed, only need to sort out the build tomorrow.  Any more comments or things I've missed let me know!

File: support-frontend/assets/pages/weekly-subscription-checkout/stage.jsx (hunk 16 above)
Comment (rupertbates) [reply support-frontend/assets/pages/weekly-subscription-checkout/stage.jsx]: Yup, I thought that too

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/paymentProviders.js (hunk 6 above)
Comment (rupertbates) [reply support-frontend/assets/pages/paper-subscription-checkout/helpers/paymentProviders.js]: Not really. As I say I would favour imports relative to the assets directory, which reminds me that you probably want to mark `assets` as a source root in intellij or it will get confused.

File: support-frontend/conf/routes (hunk 12 above)
Comment (rupertbates) [reply support-frontend/conf/routes]: See an example of a similar query param in #1701 

specifically: 
https://github.com/guardian/support-frontend/pull/1701/files#diff-ccba6c5f6c4141484e7cdca2add29512L69
and 
https://github.com/guardian/support-frontend/pull/1701/files#diff-afa248f183c4090b6973e10445c72c4aL111

File: support-frontend/assets/pages/paper-subscription-checkout/helpers/paymentProviders.js (hunk 6 above)
Comment (johnduffell) [reply support-frontend/assets/pages/paper-subscription-checkout/helpers/paymentProviders.js]: wow, this is gold, thanks, I was getting annoyed about all the imports being underlined! 🌟

File: support-frontend/conf/routes (hunk 12 above)
Comment (johnduffell) [reply support-frontend/conf/routes]: thanks. I notice that one makes you log in before doing the redirect which is what I why I ended up with two routes.

File: support-frontend/app/controllers/WeeklySubscription.scala (hunk 13 above)
Comment (rupertbates) [reply support-frontend/app/controllers/WeeklySubscription.scala]: This is fine, but do put a big TODO or something in here so we don't forget to remove it

File: support-frontend/conf/routes (hunk 12 above)
Comment (rupertbates) [reply support-frontend/conf/routes]: See comment below

**Comment by johnduffell** [comment]
thanks for sorting this one out! I will look at the prout issue
